server.py

```diff
@@ -38,8 +31,63 @@ def response_error():
                 message_in += msg
                 if len(msg) < 16:
                     break
-            conn.sendall(response_ok())
+            try:
+                parse_request(message_in)
+                conn.sendall(response_ok())
```

Comment:
My only quibble here is that attempting to send data through the socket might in itself raise an error, and it certainly wouldn't be one of the ones you've accounted for with your except clause below.  I think it'd be better to isolate the sending back of the response from the parsing of the request.